File: Learning/neo4j-rag-knowledge-base/langchain-python/rag_system.py
"""
RAG (Retrieval-Augmented Generation) System for Neo4j Knowledge Base
"""
from typing import List, Optional, Dict, Any
from langchain.chains import RetrievalQA
from langchain.prompts import PromptTemplate
from langchain.schema import Document
from langchain.chains.combine_documents import create_stuff_documents_chain
from langchain.chains import create_retrieval_chain
from langchain_core.prompts import ChatPromptTemplate
import logging

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


class Neo4jRAGSystem:
    """RAG system for querying Neo4j knowledge base"""
    
    def __init__(
        self,
        knowledge_base,
        llm,
        retrieval_k: int = 5,
        enable_hybrid_search: bool = False,
    ):
        """
        Initialize RAG system
        
        Args:
            knowledge_base: Neo4jKnowledgeBase instance
            llm: Language model instance
            retrieval_k: Number of documents to retrieve
            enable_hybrid_search: Use hybrid search instead of pure vector search
        """
        self.kb = knowledge_base
        self.llm = llm
        self.retrieval_k = retrieval_k
        self.enable_hybrid_search = enable_hybrid_search
        
        # Create custom prompt template
        self.prompt_template = self._create_prompt_template()
        
        logger.info("Initialized Neo4j RAG system")
    
    def _create_prompt_template(self) -> PromptTemplate:
        """Create prompt template for RAG"""
        template = """You are a helpful AI assistant that answers questions based on the provided context from a knowledge base.

Context from knowledge base:
{context}

Question: {question}

Instructions:
1. Answer the question based ONLY on the information provided in the context above
2. If the context doesn't contain enough information to answer the question, say "I don't have enough information in the knowledge base to answer this question"
3. Provide specific references to the source documents when possible
4. Be concise but comprehensive in your answer
5. If you're uncertain, express that uncertainty

Answer:"""
        
        return PromptTemplate(
            template=template,
            input_variables=["context", "question"]
        )
    
    def _retrieve_documents(self, query: str) -> List[Document]:
        """
        Retrieve relevant documents for query
        
        Args:
            query: User query
            
        Returns:
            List of relevant documents
        """
        if self.enable_hybrid_search:
            return self.kb.hybrid_search(query, k=self.retrieval_k)
        else:
            return self.kb.similarity_search(query, k=self.retrieval_k)
    
    def query(
        self,
        question: str,
        return_source_documents: bool = False,
        verbose: bool = False
    ) -> Dict[str, Any]:
        """
        Query the knowledge base using RAG
        
        Args:
            question: User question
            return_source_documents: Whether to return source documents
            verbose: Enable verbose output
            
        Returns:
            Dictionary with answer and optionally source documents
        """
        # Retrieve relevant documents
        documents = self._retrieve_documents(question)
        
        if not documents:
            return {
                "answer": "I couldn't find any relevant information in the knowledge base.",
                "source_documents": [] if return_source_documents else None
            }
        
        # Format context from documents
        context = self._format_context(documents)
        
        # Generate answer using LLM
        prompt = self.prompt_template.format(context=context, question=question)
        
        if verbose:
            logger.info(f"Retrieved {len(documents)} documents")
            logger.info(f"Prompt:\n{prompt}\n")
        
        answer = self.llm.invoke(prompt)
        
        result = {"answer": answer}
        
        if return_source_documents:
            result["source_documents"] = documents
        
        return result
    
    def _format_context(self, documents: List[Document]) -> str:
        """
        Format retrieved documents as context
        
        Args:
            documents: List of Document objects
            
        Returns:
            Formatted context string
        """
        context_parts = []
        
        for i, doc in enumerate(documents, 1):
            source = doc.metadata.get("source", "Unknown")
            file_name = doc.metadata.get("file_name", "Unknown")
            page = doc.metadata.get("page", "")
            
            header = f"Document {i}"
            if file_name != "Unknown":
                header += f" (Source: {file_name}"
                if page:
                    header += f", Page: {page}"
                header += ")"
            
            context_parts.append(f"{header}:\n{doc.page_content}\n")
        
        return "\n---\n".join(context_parts)
    
    def batch_query(
        self,
        questions: List[str],
        verbose: bool = False
    ) -> List[Dict[str, Any]]:
        """
        Process multiple queries in batch
        
        Args:
            questions: List of questions
            verbose: Enable verbose output
            
        Returns:
            List of results
        """
        results = []
        
        for i, question in enumerate(questions, 1):
            if verbose:
                logger.info(f"Processing query {i}/{len(questions)}: {question}")
            
            result = self.query(question, verbose=verbose)
            results.append(result)
        
        return results
    
    def chat(self, verbose: bool = False):
        """
        Interactive chat interface
        
        Args:
            verbose: Enable verbose output
        """
        print("=" * 80)
        print("Neo4j RAG Knowledge Base - Interactive Chat")
        print("=" * 80)
        print("Type 'exit' or 'quit' to end the session")
        print("Type 'sources' to see source documents with answers")
        print("=" * 80)
        print()
        
        show_sources = False
        
        while True:
            try:
                question = input("\n🤔 You: ").strip()
                
                if not question:
                    continue
                
                if question.lower() in ["exit", "quit", "q"]:
                    print("\n👋 Goodbye!")
                    break
                
                if question.lower() == "sources":
                    show_sources = not show_sources
                    status = "enabled" if show_sources else "disabled"
                    print(f"\n📚 Source documents display {status}")
                    continue
                
                print("\n🤖 Assistant: ", end="", flush=True)
                
                result = self.query(
                    question,
                    return_source_documents=show_sources,
                    verbose=verbose
                )
                
                print(result["answer"])
                
                if show_sources and result.get("source_documents"):
                    print("\n" + "=" * 80)
                    print("📚 Source Documents:")
                    print("=" * 80)
                    
                    for i, doc in enumerate(result["source_documents"], 1):
                        source = doc.metadata.get("file_name", "Unknown")
                        print(f"\n[{i}] {source}")
                        print("-" * 40)
                        preview = doc.page_content[:200] + "..." if len(doc.page_content) > 200 else doc.page_content
                        print(preview)
                
            except KeyboardInterrupt:
                print("\n\n👋 Goodbye!")
                break
            except Exception as e:
                logger.error(f"Error processing query: {e}")
                print(f"\n❌ Error: {str(e)}")


class ConversationalRAG:
    """RAG system with conversation history"""
    
    def __init__(
        self,
        knowledge_base,
        llm,
        retrieval_k: int = 5,
        max_history: int = 5,
    ):
        """
        Initialize conversational RAG
        
        Args:
            knowledge_base: Neo4jKnowledgeBase instance
            llm: Language model instance
            retrieval_k: Number of documents to retrieve
            max_history: Maximum conversation history to maintain
        """
        self.kb = knowledge_base
        self.llm = llm
        self.retrieval_k = retrieval_k
        self.max_history = max_history
        self.conversation_history: List[Dict[str, str]] = []
        
        logger.info("Initialized Conversational RAG system")
    
    def query(self, question: str) -> Dict[str, Any]:
        """
        Query with conversation context
        
        Args:
            question: User question
            
        Returns:
            Dictionary with answer
        """
        # Retrieve relevant documents
        documents = self.kb.similarity_search(question, k=self.retrieval_k)
        
        # Build context with history
        context = self._build_context_with_history(documents)
        
        # Create prompt with history
        prompt = self._create_conversational_prompt(context, question)
        
        # Generate answer
        answer = self.llm.invoke(prompt)
        
        # Update conversation history
        self.conversation_history.append({
            "question": question,
            "answer": answer
        })
        
        # Trim history
        if len(self.conversation_history) > self.max_history:
            self.conversation_history = self.conversation_history[-self.max_history:]
        
        return {
            "answer": answer,
            "source_documents": documents
        }
    
    def _build_context_with_history(self, documents: List[Document]) -> str:
        """Build context including conversation history"""
        context_parts = []
        
        # Add document context
        for i, doc in enumerate(documents, 1):
            context_parts.append(f"Document {i}:\n{doc.page_content}\n")
        
        return "\n---\n".join(context_parts)
    
    def _create_conversational_prompt(self, context: str, question: str) -> str:
        """Create prompt with conversation history"""
        prompt_parts = [
            "You are a helpful AI assistant. Answer questions based on the provided context and conversation history.",
            "\nContext from knowledge base:",
            context,
        ]
        
        if self.conversation_history:
            prompt_parts.append("\nConversation History:")
            for i, exchange in enumerate(self.conversation_history[-3:], 1):
                prompt_parts.append(f"\nQ{i}: {exchange['question']}")
                prompt_parts.append(f"A{i}: {exchange['answer']}")
        
        prompt_parts.append(f"\nCurrent Question: {question}")
        prompt_parts.append("\nAnswer:")
        
        return "\n".join(prompt_parts)
    
    def clear_history(self):
        """Clear conversation history"""
        self.conversation_history = []
        logger.info("Cleared conversation history")

File: Learning/neo4j-rag-knowledge-base/vscode-extension/src/ragSystem.ts
/**
 * RAG System with GitHub Copilot Chat integration
 */
import * as vscode from 'vscode';
import { Neo4jKnowledgeBase, Document } from './neo4jKb';

export class RAGSystem {
    private kb: Neo4jKnowledgeBase;
    private retrievalK: number;

    constructor(kb: Neo4jKnowledgeBase, retrievalK: number) {
        this.kb = kb;
        this.retrievalK = retrievalK;
    }

    public async query(question: string): Promise<{ answer: string; sources: Document[] }> {
        // Retrieve relevant documents
        const documents = await this.kb.similaritySearch(question, this.retrievalK);

        if (documents.length === 0) {
            return {
                answer: "I couldn't find any relevant information in the knowledge base.",
                sources: []
            };
        }

        // Format context for Copilot
        const context = this.formatContext(documents);

        // Use GitHub Copilot Chat API to generate answer
        const answer = await this.generateAnswerWithCopilot(question, context);

        return {
            answer: answer,
            sources: documents
        };
    }

    private formatContext(documents: Document[]): string {
        const contextParts: string[] = [];

        documents.forEach((doc, index) => {
            const header = `Document ${index + 1} (Source: ${doc.metadata.fileName || 'Unknown'})`;
            contextParts.push(`${header}:\n${doc.content}\n`);
        });

        return contextParts.join('\n---\n');
    }

    private async generateAnswerWithCopilot(question: string, context: string): Promise<string> {
        // Create a prompt for GitHub Copilot
        const prompt = `You are a helpful AI assistant that answers questions based on the provided context from a knowledge base.

Context from knowledge base:
${context}

Question: ${question}

Instructions:
1. Answer the question based ONLY on the information provided in the context above
2. If the context doesn't contain enough information, say "I don't have enough information in the knowledge base to answer this question"
3. Provide specific references to the source documents when possible
4. Be concise but comprehensive in your answer

Answer:`;

        try {
            // Use VS Code language model API with GitHub Copilot
            const models = await vscode.lm.selectChatModels({
                vendor: 'copilot',
                family: 'gpt-4'
            });

            if (models.length === 0) {
                return "GitHub Copilot is not available. Please ensure GitHub Copilot Pro is enabled.";
            }

            const model = models[0];
            const messages = [
                vscode.LanguageModelChatMessage.User(prompt)
            ];

            const response = await model.sendRequest(messages, {}, new vscode.CancellationTokenSource().token);

            let answer = '';
            for await (const chunk of response.text) {
                answer += chunk;
            }

            return answer.trim();
        } catch (error) {
            return `Error generating answer with Copilot: ${error}`;
        }
    }

    public async openChatPanel(): Promise<void> {
        const panel = vscode.window.createWebviewPanel(
            'neo4jRagChat',
            'Neo4j RAG Chat',
            vscode.ViewColumn.Two,
            {
                enableScripts: true,
                retainContextWhenHidden: true
            }
        );

        panel.webview.html = this.getChatWebviewContent();

        // Handle messages from the webview
        panel.webview.onDidReceiveMessage(
            async message => {
                switch (message.command) {
                    case 'query':
                        const result = await this.query(message.text);
                        panel.webview.postMessage({
                            command: 'response',
                            answer: result.answer,
                            sources: result.sources
                        });
                        break;
                }
            },
            undefined,
            []
        );
    }

    private getChatWebviewContent(): string {
        return `<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Neo4j RAG Chat</title>
    <style>
        body {
            font-family: var(--vscode-font-family);
            color: var(--vscode-foreground);
            background-color: var(--vscode-editor-background);
            padding: 20px;
            margin: 0;
        }
        #chat-container {
            max-width: 800px;
            margin: 0 auto;
        }
        #messages {
            height: 500px;
            overflow-y: auto;
            border: 1px solid var(--vscode-panel-border);
            padding: 15px;
            margin-bottom: 20px;
            border-radius: 5px;
        }
        .message {
            margin-bottom: 15px;
            padding: 10px;
            border-radius: 5px;
        }
        .user-message {
            background-color: var(--vscode-input-background);
            text-align: right;
        }
        .assistant-message {
            background-color: var(--vscode-editor-inactiveSelectionBackground);
        }
        .sources {
            margin-top: 10px;
            padding: 10px;
            background-color: var(--vscode-textBlockQuote-background);
            border-left: 3px solid var(--vscode-textBlockQuote-border);
            font-size: 0.9em;
        }
        #input-container {
            display: flex;
            gap: 10px;
        }
        #query-input {
            flex: 1;
            padding: 10px;
            background-color: var(--vscode-input-background);
            color: var(--vscode-input-foreground);
            border: 1px solid var(--vscode-input-border);
            border-radius: 3px;
        }
        #send-button {
            padding: 10px 20px;
            background-color: var(--vscode-button-background);
            color: var(--vscode-button-foreground);
            border: none;
            border-radius: 3px;
            cursor: pointer;
        }
        #send-button:hover {
            background-color: var(--vscode-button-hoverBackground);
        }
    </style>
</head>
<body>
    <div id="chat-container">
        <h1>Neo4j RAG Knowledge Base Chat</h1>
        <div id="messages"></div>
        <div id="input-container">
            <input type="text" id="query-input" placeholder="Ask a question..." />
            <button id="send-button">Send</button>
        </div>
    </div>

    <script>
        const vscode = acquireVsCodeApi();
        const messagesDiv = document.getElementById('messages');
        const queryInput = document.getElementById('query-input');
        const sendButton = document.getElementById('send-button');

        function addMessage(text, isUser) {
            const messageDiv = document.createElement('div');
            messageDiv.className = 'message ' + (isUser ? 'user-message' : 'assistant-message');
            messageDiv.textContent = text;
            messagesDiv.appendChild(messageDiv);
            messagesDiv.scrollTop = messagesDiv.scrollHeight;
        }

        function addSources(sources) {
            if (sources && sources.length > 0) {
                const sourcesDiv = document.createElement('div');
                sourcesDiv.className = 'sources';
                sourcesDiv.innerHTML = '<strong>Sources:</strong><br>' + 
                    sources.map((s, i) => \`[\${i+1}] \${s.metadata.fileName || 'Unknown'}\`).join('<br>');
                messagesDiv.appendChild(sourcesDiv);
                messagesDiv.scrollTop = messagesDiv.scrollHeight;
            }
        }

        function sendQuery() {
            const query = queryInput.value.trim();
            if (query) {
                addMessage(query, true);
                vscode.postMessage({ command: 'query', text: query });
                queryInput.value = '';
            }
        }

        sendButton.addEventListener('click', sendQuery);
        queryInput.addEventListener('keypress', (e) => {
            if (e.key === 'Enter') {
                sendQuery();
            }
        });

        window.addEventListener('message', event => {
            const message = event.data;
            if (message.command === 'response') {
                addMessage(message.answer, false);
                addSources(message.sources);
            }
        });
    </script>
</body>
</html>`;
    }
}
